core/commands/pin.go

```diff
@@ -414,6 +436,90 @@ func pinLsAll(typeStr string, ctx context.Context, n *core.IpfsNode) (map[string
 	return keys, nil
 }
 
+type pinStatus struct {
+	// use pointer to array slice to reduce memory usage
```

Comment:
A slice is the width of three pointers.  There is a node for every indirect pin.  By using an extra lawyer of indirection I reduce the size by two pointers for each node.  The actual array slice will be shared many times in the common case.